src/charm.py

```diff
@@ -98,8 +98,13 @@ def on_pebble_ready(self, event: EventBase) -> None:
             event: The event object of the pebble ready event
         """
         container = event.workload
-        config = self._config_file(event)
-        container.push(CONFIG_PATH, config)
+        if not self._stored.agent_config:
```

Comment:
Could you please add a comment for what this `if` is checking? I don't understand why the old code is wrong.

Also, recently @mmanciop (?) encountered scenarios where stored wasn't reliable for this purpose here so now we're moving to storing in peer relation app data (if it's the same stuff for all units; you cannot use peer unit data because in the case of a single unit you won't have that bag).

Reply:
The config file is built on pebble_ready and completed once you relate grafana agent with loki.

I'm saving the configuration in stored because if you have a grafana agent and loki charms related, when you reboot the VM in which microk8s is running, the `loki_push_api_endpoint_joined` is not fired again and the config will be in a different state (without loki section)

Reply:
The old code,
```python
config = self._config_file(event)
container.push(CONFIG_PATH, config)
```
seems to be doing the right thing already, but I guess I'm missing something here.
Ideally the full config should be constructible upon re-entry without stored state. Would that be possible here?

Reply:
Arguably, we may want to see whether the config is actually different, so it's not constantly interrupting the binary and the flow of logs

Reply:
Well we are using the reload web api, I don’t think it is a big disruption

Reply:
@rbarry82 before this PR, when you reboot the VM in which microk8s is running the config ends up with this:

```
...
loki: {}
prometheus:
  configs:
  - name: agent_scraper
    remote_write: []
    scrape_configs: []
server:
  log_level: info
  ```
But should be:

```
...
loki:
  configs:
  - clients:
    - url: http://loki-k8s-0.loki-k8s-endpoints.observability.svc.cluster.local:3100/loki/api/v1/push
    name: promtail
    positions:
      filename: /tmp/positions.yaml
    scrape_configs:
    - job_name: loki
      loki_push_api:
        server:
          grpc_listen_port: 3600
          http_listen_port: 3500
prometheus:
  configs:
  - name: agent_scraper
    remote_write: []
    scrape_configs: []
server:
  log_level: info
```

Reply:
> Well we are using the reload web api, I don’t think it is a big disruption

@mmanciop we wanted to use the reload web api, but we [found a bug](https://github.com/canonical/grafana-agent-k8s-operator/issues/19):

```python
# FIXME: #19
# self._reload_config()
self._container.restart(self._name)
```

Reply:
@Abuelodelanada Granted, I was making a comment on "we may not want to rebuild the config and restart it on every config change -- only if it's different", which is a good reason to either pull it from the container and compare or to keep it stored off somewhere.

Reply:
@rbarry82 we are doing this. The lines you are reading (101-104) are executed `on_pebble_ready` method that is executed only while deploying the unit.

I think we are doing what are you proposing:

```python
        config = self._config_file(event)

        try:
            old_config = self._container.pull(CONFIG_PATH)
        except PathError:
            # If the file does not yet exist, pebble_ready has not run yet
            pass

        try:
            if yaml.safe_load(config) != yaml.safe_load(old_config):
                self._container.push(CONFIG_PATH, config)
                self._stored.agent_config = config
                # FIXME: #19
                # self._reload_config()
                self._container.restart(self._name)
```

Or I'm missing something?